pkg/aws/ec2/ec2_test.go

```diff
@@ -379,3 +300,197 @@ func TestCollector_Collect(t *testing.T) {
+				RunAndReturn(tt.GetProducts).
+				Times(tt.expectedCalls)
+
+			err := collector.populateStoragePricingMap(tt.ctx)
+			if tt.err != nil {
+				assert.ErrorIs(t, err, tt.err)
+			}
+			assert.Equal(t, tt.expected, collector.storagePricingMap.Regions)
+		})
+	}
+}
+
+func Test_FetchVolumesData(t *testing.T) {
+	t.Run("sends EBS volumes data to channel", func(t *testing.T) {
+		regionName := "af-south-1"
+		region := ec2Types.Region{
+			RegionName: aws.String(regionName),
+		}
+
+		ps := mockpricing.NewPricing(t)
+		collector := New(&Config{
+			Regions: []ec2Types.Region{region},
+			Logger:  logger,
+		}, ps)
+
+		client := mockec2.NewEC2(t)
+		client.EXPECT().
+			DescribeVolumes(mock.Anything, mock.Anything, mock.Anything).
+			RunAndReturn(
+				func(ctx context.Context, e *ec2.DescribeVolumesInput, optFns ...func(*ec2.Options)) (*ec2.DescribeVolumesOutput, error) {
+					return &ec2.DescribeVolumesOutput{
+						Volumes: []ec2Types.Volume{
+							{
+								VolumeId: aws.String("vol-111111111"),
+							},
+						},
+					}, nil
+				},
+			).
+			Times(1)
+
+		wg := sync.WaitGroup{}
+		wg.Add(len(collector.Regions))
+		ch := make(chan []ec2Types.Volume)
+		go collector.fetchVolumesData(context.Background(), client, regionName, ch)
+		go func() {
+			wg.Wait()
+			close(ch)
+		}()
+
+		msg, ok := <-ch
+		assert.True(t, ok)
+		assert.NotNil(t, msg)
+		assert.IsType(t, []ec2Types.Volume{}, msg)
+	})
+}
+
+func Test_EmitMetricsFromVolumesChannel(t *testing.T) {
+	t.Run("reads from volumes channel and sends it over to prometheus channel", func(t *testing.T) {
+		volumesCh := make(chan []ec2Types.Volume)
+		promCh := make(chan prometheus.Metric)
+
+		regionName := "af-south-1"
+		region := ec2Types.Region{
+			RegionName: aws.String(regionName),
+		}
+		volumeType := "gp3"
+
+		ps := mockpricing.NewPricing(t)
+		collector := New(&Config{
+			Regions: []ec2Types.Region{region},
+			Logger:  logger,
+		}, ps)
+
+		collector.storagePricingMap = &StoragePricingMap{
+			Regions: map[string]*StoragePricing{
+				regionName: {
+					Storage: map[string]float64{
+						volumeType: 0.1047,
+					},
+				},
+			},
+		}
+
+		originMsg := []ec2Types.Volume{
+			{
+				AvailabilityZone: aws.String(fmt.Sprintf("%sa", regionName)),
+				VolumeId:         aws.String("vol-111111111"),
+				VolumeType:       ec2Types.VolumeType(volumeType),
+				Size:             aws.Int32(100),
+			},
+		}
+
+		// fill volumes channel with data from the above volume
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			volumesCh <- originMsg
+			wg.Wait()
+			close(volumesCh)
+		}()
+
+		wg2 := sync.WaitGroup{}
+		wg2.Add(1)
+		go func() {
+			defer wg2.Done()
+			collector.emitMetricsFromVolumesChannel(volumesCh, promCh)
+			wg2.Wait()
+		}()
+
+		receivedMsg, ok := <-promCh
+		close(promCh)
+
+		assert.True(t, ok)
+		assert.NotNil(t, receivedMsg)
+		assert.Contains(t, receivedMsg.Desc().String(), "persistent_volume_usd_per_hour")
```

Comment:
slightly more idiomatic maybe?

```suggestion
		wg := sync.WaitGroup{}
		wg.Add(2)
		var recMsg prometheus.Metric
		var ok bool

		// fill volumes channel with data from the above volume
		go func() {
			defer wg.Done()
			defer close(volumesCh)
			volumesCh <- originMsg
		}()

		go func() {
			defer wg.Done()
			recMsg, ok = <-promCh
		}()

		go collector.emitMetricsFromVolumesChannel(volumesCh, promCh)

		wg.Wait()

		assert.True(t, ok)
		assert.NotNil(t, recMsg)
		assert.Contains(t, recMsg.Desc().String(), "persistent_volume_usd_per_hour")
```

Reply:
Basically, we need these three processes to happen in parallel:

- emitMetrics
- filling the `volumeCh`
- reading from the `promCh`

Essentially, `emitMetrics` blocks until the `volumeCh` has something it can read from, `promCh` is populated by `emitMetrics`, and `volumeCh` needs to _close_ after sending data so that `emitMetrics` knows there's no more values to loop over.

So:
- `emitMetrics` will only return when `volumeCh` is closed
- `recMsg` can only be evaluated after it has been read from `promCh` and _after_ emitMetrics returns

the three things done can actually be written in any order, `go` will magically make them do what we think.  We just need to have a `waitGroup` there to ensure that the thing that initiates the "termination" of the process (closing the `volumeCh`) blocks the `assert`, hence the WaitGroup.

Reply:
We also don't _need_ to close the promCh here, but you can.

> Another note: Channels aren't like files; you don't usually need to close them. Closing is only necessary when the receiver must be told there are no more values coming, such as to terminate a range loop.

ref: https://go.dev/tour/concurrency/4

Reply:
You can even do something like:

```go
                originMsgs := []ec2Types.Volume{
			{
				AvailabilityZone: aws.String(fmt.Sprintf("%sa", regionName)),
				VolumeId:         aws.String("vol-111111111"),
				VolumeType:       ec2Types.VolumeType(volumeType),
				Size:             aws.Int32(100),
			},
			{
				AvailabilityZone: aws.String(fmt.Sprintf("%sa", regionName)),
				VolumeId:         aws.String("vol-111111111"),
				VolumeType:       ec2Types.VolumeType(volumeType),
				Size:             aws.Int32(100),
			},
		}

		wg := sync.WaitGroup{}
		wg.Add(1)

		// fill volumes channel with data from the above volume
		go func() {
			defer wg.Done()
			defer close(volumesCh)
			volumesCh <- originMsgs
		}()
		go collector.emitMetricsFromVolumesChannel(volumesCh, promCh)

		wg.Wait()

		for i := 0; i < len(originMsgs); i++ {
			m, ok := <-promCh

			assert.True(t, ok)
			assert.NotNil(t, m)
			assert.Contains(t, m.Desc().String(), "persistent_volume_usd_per_hour")
		}
```

Note that if you loop over the promCh _the amount of expected times_, this will work.  But if you change it to looping over the channel itself:

```go
		for m := range promCh {
			// m, ok := <-promCh

			// assert.True(t, ok)
			assert.NotNil(t, m)
			assert.Contains(t, m.Desc().String(), "persistent_volume_usd_per_hour")
		}
```

you'll get into trouble, because that channel will never be closed, and the test will hang.

Reply:
And now that I look at it even more, you can even just do this:

```go
		collector.storagePricingMap = &StoragePricingMap{
			Regions: map[string]*StoragePricing{
				regionName: {
					Storage: map[string]float64{
						volumeType: 0.1047,
					},
				},
			},
		}

		originMsgs := []ec2Types.Volume{
			{
				AvailabilityZone: aws.String(fmt.Sprintf("%sa", regionName)),
				VolumeId:         aws.String("vol-111111111"),
				VolumeType:       ec2Types.VolumeType(volumeType),
				Size:             aws.Int32(100),
			},
		}

		go collector.emitMetricsFromVolumesChannel(volumesCh, promCh)
		volumesCh <- originMsgs
		close(volumesCh)

		m, ok := <-promCh
		assert.True(t, ok)
		assert.NotNil(t, m)
		assert.Contains(t, m.Desc().String(), "persistent_volume_usd_per_hour")
```

All that's required is that the `emitMetrics` function is in the background such that it can receive on the `volumesCh`, and emit on the `promCh`.

`promCh` will block on receiving,  and closing the `volumesCh` indicates to `emitMetrics` that there's nothing more to loop over.

In fact, if you don't close the `volumesCh`, the test will still pass, but `emitMetrics` will have never returned (which I guess we don't care about for the purposes of this test)